### PR TITLE
add Persian data

### PR DESCRIPTION
**Pull request by saeiddrv**

### CODE REVIEW

**Comment by ojwb** [comment]
I've merged master to the PR branch so I can do the same for the snowball PR (otherwise CI tests will fail for any existing algorithms which have been added or changed since you branched).

**Comment by saeiddrv** [comment]
Thank you for merging master into the branch. I've now updated the output.txt to reflect the latest algorithm improvements and pushed the changes.

File: persian/COPYING [hunk 1]
```diff
@@ -0,0 +1,7 @@
+voc.txt is generated by:
+
+https://github.com/behnam/persian-words-frequency
```
Comment (ojwb) [inline persian/COPYING]: `voc.txt` seems to correspond to all the entries from `persian-wikipedia.txt` in that repo, except that a single entry consisting only of U+200C has been removed.

That gives us 406159 words, but 213661 of those have a frequency of occurrence of just 1 in the original list.  We've used similar analysis of wikipedia dumps to get word lists for other languages, and have found that the very low occurrence entries are generally better removed - some are real words, but a lot are typos and other junk.  It looks like the lowest threshold we've used was 7 for indonesian (which had a relatively low number of pages on wikipedia), but otherwise it's at least 10.

`voc.txt` doesn't need to be perfectly clean of junk terms, words from other languages, etc as the stemmer will encounter them in real use so testing handling of them is reasonable, but these lists serve a dual purpose - they are used to test that the stemmer gives expected results across target languages, but also to provide a vocabulary for evaluating proposed changes to the stemmer.  An excess of junk entries is less of a problem for the first use (though it makes running the tests take longer, which becomes a problem for development as the number of stemmers we support grows).  For the latter use, an excess of entries which aren't valid Persian words (or are valid but extremely rare) is actively unhelpful.

Also, while U+200C has been removed, e.g. entries consisting only of a single U+200D and U+2010 remain, as do entries consisting only of a string of U+200C and/or U+200D - I counted 19 in all.  I think it probably makes sense to only keep entries with at least one letter.  I can easily fix that if you agree.

I also notice the list is 15 years old so must be based on wikipedia data at least that old and there's probably significantly more text in wikipedia now.

I fetched the latest fawiki dump and ran:

```
$ scripts/wikipedia-dump-to-freq fawiki-latest-pages-articles.xml.bz2 1 '[\p{Script:arabic}\x{2010}\x{200c}\x{200d}]' > fa.freq
```

(This selects words consisting of Unicode characters in the "arabic" script plus U+2010, U+200C and U+200D.)

It gives 1176187 entries, 656691 occurring >= 2 times, 482998 >= 3 times, 380787 >= 4 times - `persian-wikipedia.txt` is somewhere between those last two in size.

Even if we required 10 occurrences, we'd still have 182703 entries.

(21 entries don't contain a single letter, very similar to the 20 in `persian-wikipedia.txt`.)

Should I put the output somewhere for you to take a look?  It's ~21MB.

Comment (saeiddrv) [reply persian/COPYING]: Yes please, that would be great. Thanks!

Comment (ojwb) [reply persian/COPYING]: Here it is gzipped:

[fa.freq.gz](https://github.com/user-attachments/files/26691193/fa.freq.gz)

Format is a line for each word with format: `<frequency><tab character><word>`

If you want a `voc.txt` for testing, `scripts/freq-to-voc` in `snowball-data` may be useful - it can optionally select only words with at least a specified frequency, and also produces a sorted output.

Looking at this frequency list, if we sum occurrences for words occurring 10 or more times we get 102,012,222 total occurrences (and 2,295,147 total occurrences for all words occurring < 10 times).  That means a word occurring 9 times in this list is only one in `(102012222+2295147)/9` or about 1 in 11.6 million words in the source text.

The appropriate vocabulary size here depends on the language.  We're including inflected/derived forms and that varies greatly by language, as does the number of headwords in a typical working vocabulary for a language.

So I'm not really sure what size vocabulary list is appropriate for Persian, but 400 thousand seems much too high.  Only Tamil and Arabic have more than this of the current languages (Arabic's current list is essentially a mechanical generation of all possible forms, which is not a helpful approach and we shouldn't use that as any sort of guide for Persian - see #31 for details; I suspect Tamil's is also too large).

We ideally want to cover all the prefixes/suffixes the stemmer handles, and beyond that to have enough of a sample of the working vocabulary to usefully evaluate a proposed change.  A lot of rare words can be counterproductive when evaluating changes though - we still probably want to make a change that really helps common words even if it negatively affects valid words which are very rarely used in practice.

A threshold of 10 would give 182,703 entries; a threshold of 20 gives 109,523 entries; a threshold of 30 gives 82,390 entries.  I suspect a threshold somewhere in this range is where we want to be, but if you could take a look at the entries it may be clear where we transition from mostly real words which are useful to consider to mostly non-words/non-Persian words/very rare Persian words.

(These numbers are before removing all-whitespace entries, but there's only 21 on them.)

Comment (saeiddrv) [reply persian/COPYING]: I regenerated `voc.txt` from the frequency list you shared — used a threshold of ≥20 occurrences and filtered out entries without actual Persian letters (digits, whitespace-only, etc.). That gives us 104,267 words, down from the original 406K. Pushed in 7d5ba89.

Comment (ojwb) [reply persian/COPYING]: I try to record a command which generates `voc.txt` (at least for cases where the processing is automated) as it can be useful in the future if we discover a problem with the automated processing, or if we just want to process a newer version of the source data in the same way.

Based on your description I tried:

```
gzip -dc fa.freq.gz|scripts/freq-to-voc 20 | grep '[[:alpha:]]' > persian/voc.txt
```

That selects everything which occurs at least 20 times and then removes anything with a letter in it (and I selected `arabic` script when I generated `fa.freq` so the only letters should be Arabic ones).

I get all the entries you did, but also 5249 extras.  With a threshold of 21 I'm missing some entries, so 20 seems correct.  I guess you are filtering to just the Arabic characters used in Persian?

Comment (saeiddrv) [reply persian/COPYING]: Yes, I applied stricter filtering — removed entries containing any digits (Persian ۰-۹, Arabic ٠-٩, ASCII 0-9), entries with fewer than 2 Arabic-script letters, and entries starting with zero-width characters. Those 5,249 extras are mostly malformed words like `۹۵خانوار` or standalone numbers — they're not valid Persian words, just Wikipedia artifacts.

Comment (ojwb) [reply persian/COPYING]: > Those 5,249 extras are mostly malformed words like ۹۵خانوار or standalone numbers — they're not valid Persian words, just Wikipedia artifacts.

Sure - I'm not arguing we want these additional entries - as I said, I'm trying to document the steps to generate this list.

Based on your description I can get much closer (the list only contains Arabic script characters so I can just check for any digits and letters):

```
gzip -dc fa.freq.gz | scripts/freq-to-voc 20 | perl -CS -ne 'print if !/\d/ and /^\p{Letter}.*\p{Letter}/' > persian/voc.txt
```

However I still get one extra entry compared to you, which is exactly two U+0622 characters.  That's a character the stemmer recognises, and seems to be an Arabic letter:

```
0622;ARABIC LETTER ALEF WITH MADDA ABOVE;Lo;0;AL;0627 0653;;;;N;ARABIC LETTER MADDAH ON ALEF;;;;
```

This doesn't seem to be a valid word, but doesn't seem to be excluded by your criteria - did you just manually remove it?  Or was there another criterion?  (It's not length > 2 as there are 14 other two letter words.)

Comment (ojwb) [reply persian/COPYING]: Another very minor point is we lack any entries containing `{ar_heh}`, which is the only `among` case not exercised now if we ignore the one for ASCII space.  The original `voc.txt` also lacked an instance of this.

The entries in `fa.freq` containing `{ar_heh}` with a frequency of at least 3 are:

```
11	بہ
10	شدہ
9	کردہ
8	شاہ
8	ماہ
8	ہے
7	کہ
5	جنازہ
4	مغازہ
4	گاہ
4	اللہ
3	شیدہ
3	دادہ
3	تذکرہ
3	دانشگاہ
3	ھمراہ
3	دادگاہ
3	کلهورہ
3	راہ
3	ہیں
3	رہی
```

Can you pick one (or suggest a suitable example word if none of these are real Persian words)?  We can add it as a hand-picked addition.

To summarise the current status, I think it's just these two points to resolve before we merge:

* Was the removal of `{aa}{aa}` based on a rule or just an ad-hoc hand edit?  Either is OK, but if there's a rule (which happens to only remove one additional entry for this dataset) it would be useful to know what it was for future reuse.
* Add a word containing `{ar_heh}` to improve code coverage.

Comment (saeiddrv) [reply persian/COPYING]: For `آآ` — it's not a valid Persian word, I removed it manually.

For `{ar_heh}` — I'll add `دانشگاہ` to `voc.txt` with expected output `دانش`. It tests both the U+06C1 → U+0647 normalization and the `-gah` suffix stripping.
